src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
@IlyaBiryukov StreamJsonRpc doesn't call `FailFast`. Roslyn is the one calling `JsonRpc.Dispose()` while it is still making outbound calls to it. So the exception thrown from `AsyncSemaphore` is propagating through `JsonRpc` to its caller, Roslyn, who decides to crash when the situation they created bites them. 
Consider: wouldn't `JsonRpc` be justified in throwing `ObjectDisposedException` itself (independent of `AsyncSempahore`) if it were called to do something after it was disposed, or if it were disposed while still performing an operation for another context? I think so. If you buy that, then the fix still needs to go to Roslyn.